crates/conjure_core/src/rule_engine/rewrite.rs

```diff
@@ -431,3 +436,10 @@ fn check_priority<'a>(
         \n Rules{:?}", initial_expr, new_expr, rules)
     }
 }
+fn trace_rules(results: &[RuleResult], expression: Expression) {
+    if !results.is_empty() {
+        let rule = results[0].rule;
+        let new_expression = results[0].reduction.new_expression.clone();
+        trace!(target: "rule_engine", "Rule applied: {:?}, to Expression: {:?}, resulting in: {:?}", rule.name, expression, new_expression);
```

Comment:
```suggestion
        trace!(target: "rule_engine", "Rule applied: {:?}, to Expression: {}, resulting in: {}", rule.name, expression, new_expression);
```

Expression implements Display, so we don't need to debug print here.
I would take a look at what we do on main here.